File: ws01_plumbing/src/cpp01_topic/src/demo02_talker_stu.cpp

#include "rclcpp/rclcpp.hpp"

#include "base_interfaces_demo/msg/student.hpp"

using base_interfaces_demo::msg::Student;
using namespace std::chrono_literals;

class TalkerStu: public rclcpp::Node{
public:
    TalkerStu():Node("talker_stu"){

        // 创建发布方
        publisher_ = this->create_publisher<Student>("chatter_stu", 10);
        // 创建定时器
        timer_ = this->create_wall_timer(500ms, std::bind(&TalkerStu::on_timer, this));

    }

private:
    void on_timer(){
        auto stu = Student();
        stu.name = "fujun";
        stu.age = 26;
        stu.height = 175.0;
        publisher_->publish(stu);
        RCLCPP_INFO(this->get_logger(), "发布的消息: (%s,%d,%f)", stu.name.c_str(), stu.age, stu.height);
    }

    rclcpp::Publisher<Student>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};


int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TalkerStu>());
    rclcpp::shutdown();
    return 0;
}

File: ws01_plumbing/src/py01_topic/py01_topic/demo02_talker_stu.py


import rclpy
from rclpy.node import Node
from base_interfaces_demo.msg import Student

class TalkerStu(Node):
    def __init__(self):
        super().__init__("talker_stu_py")
        self.get_logger().info("创建发布者 Python")
        self.publisher = self.create_publisher(Student, "chatter_stu", 10)
        self.timer = self.create_timer(0.5, self.on_timer)

    def on_timer(self):
        msg = Student()
        msg.name = "hahaha"
        msg.age = 20
        msg.height = 21.5
        self.publisher.publish(msg)


def main():
    rclpy.init()
    rclpy.spin(TalkerStu())
    rclpy.shutdown()


if __name__ == '__main__':
    main()